File: PregnancyWeight.java
// PregnancyWeight.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

class Entry {
    String date;
    double weight;
    String note;
    double bmi;
}

class Data {
    double height;
    double preWeight;
    List<Entry> entries = new ArrayList<>();
}

public class PregnancyWeight {
    private Data data = new Data();
    private final String dataFile = "pregnancy_weight.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public PregnancyWeight() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                data = gson.fromJson(json, Data.class);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(data).getBytes());
        } catch (Exception e) {}
    }

    public void setHeight(double cm) {
        data.height = cm;
        save();
        System.out.printf("✅ Height set to %.1f cm%n", cm);
    }

    public void setPreWeight(double kg) {
        data.preWeight = kg;
        save();
        System.out.printf("✅ Pre‑pregnancy weight set to %.1f kg%n", kg);
    }

    private double bmi(double weight) {
        if (data.height <= 0) return 0;
        double h = data.height / 100.0;
        return weight / (h * h);
    }

    private String bmiCategory(double bmiVal) {
        if (bmiVal < 18.5) return "Underweight";
        if (bmiVal < 25.0) return "Normal";
        if (bmiVal < 30.0) return "Overweight";
        return "Obese";
    }

    public void add(double weight, String dateStr, String note) {
        if (data.height <= 0) {
            System.out.println("❌ Please set your height first: height <cm>");
            return;
        }
        if (dateStr == null) dateStr = LocalDate.now().toString();
        Entry entry = new Entry();
        entry.date = dateStr;
        entry.weight = weight;
        entry.note = note != null ? note : "";
        entry.bmi = bmi(weight);
        data.entries.add(entry);
        data.entries.sort((a, b) -> a.date.compareTo(b.date));
        save();
        System.out.printf("✅ Logged %.1f kg on %s%n", weight, dateStr);
    }

    public void list() {
        if (data.entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        System.out.println("\n📋 Weight Entries:");
        for (Entry e : data.entries) {
            String note = e.note.isEmpty() ? "" : " – " + e.note;
            String bmiStr = e.bmi > 0 ? String.format(" (BMI: %.1f)", e.bmi) : "";
            System.out.printf("  %s: %.1f kg%s%s%n", e.date, e.weight, bmiStr, note);
        }
    }

    public void stats() {
        if (data.entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        List<Double> weights = new ArrayList<>();
        for (Entry e : data.entries) weights.add(e.weight);
        int total = weights.size();
        double current = weights.get(total - 1);
        double avg = weights.stream().mapToDouble(Double::doubleValue).average().orElse(0);
        double mn = weights.stream().mapToDouble(Double::doubleValue).min().orElse(0);
        double mx = weights.stream().mapToDouble(Double::doubleValue).max().orElse(0);
        Double gain = data.preWeight > 0 ? current - data.preWeight : null;

        System.out.println("\n📊 Statistics:");
        System.out.printf("  Entries: %d%n", total);
        System.out.printf("  Current: %.1f kg%n", current);
        System.out.printf("  Average: %.1f kg%n", avg);
        System.out.printf("  Min: %.1f kg%n", mn);
        System.out.printf("  Max: %.1f kg%n", mx);
        if (gain != null) System.out.printf("  Total gain: %+.1f kg%n", gain);
    }

    public void showBMI() {
        if (data.entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        if (data.height <= 0) {
            System.out.println("❌ Height not set.");
            return;
        }
        Entry current = data.entries.get(data.entries.size() - 1);
        double bmiVal = bmi(current.weight);
        String cat = bmiCategory(bmiVal);
        System.out.printf("\n📐 Current BMI: %.1f (%s)%n", bmiVal, cat);
        System.out.println("BMI Range: 18.5 - 24.9");
    }

    public void showGain() {
        if (data.entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        if (data.preWeight <= 0) {
            System.out.println("❌ Pre‑pregnancy weight not set. Use: preweight <kg>");
            return;
        }
        Entry current = data.entries.get(data.entries.size() - 1);
        Entry first = data.entries.get(0);
        double gain = current.weight - data.preWeight;
        System.out.println("\n📈 Weight Gain:");
        System.out.printf("  Pre‑pregnancy: %.1f kg%n", data.preWeight);
        System.out.printf("  First logged: %.1f kg%n", first.weight);
        System.out.printf("  Current: %.1f kg%n", current.weight);
        System.out.printf("  Total gain: %+.1f kg%n", gain);
    }

    public void trend(int days) {
        if (data.entries.size() < days) {
            System.out.printf("Need at least %d entries for trend.%n", days);
            return;
        }
        List<Entry> recent = data.entries.subList(data.entries.size() - days, data.entries.size());
        double avg = recent.stream().mapToDouble(e -> e.weight).average().orElse(0);
        System.out.printf("\n📈 Trend (last %d days):%n", days);
        System.out.printf("  Average weight: %.1f kg%n", avg);
        System.out.println("  Recent entries:");
        for (int i = Math.max(0, recent.size() - 5); i < recent.size(); i++) {
            Entry e = recent.get(i);
            System.out.printf("    %s: %.1f kg%n", e.date, e.weight);
        }
    }

    public void export(String filename) throws Exception {
        if (data.entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        try (BufferedWriter writer = Files.newBufferedWriter(Paths.get(filename))) {
            writer.write("Date,Weight (kg),BMI,Note\n");
            for (Entry e : data.entries) {
                String bmiStr = e.bmi > 0 ? String.format("%.1f", e.bmi) : "";
                writer.write(String.format("%s,%.1f,%s,%s%n", e.date, e.weight, bmiStr, e.note));
            }
        }
        System.out.printf("✅ Exported %d entries to %s%n", data.entries.size(), filename);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: PregnancyWeight <command> [options]");
            return;
        }
        PregnancyWeight app = new PregnancyWeight();
        String cmd = args[0];
        Map<String, String> params = new HashMap<>();
        for (int i=1; i<args.length; i++) {
            if (args[i].startsWith("--") && i+1 < args.length) {
                params.put(args[i].substring(2), args[++i]);
            } else if (!args[i].startsWith("--")) {
                params.put("arg", args[i]);
            }
        }
        switch (cmd) {
            case "height":
                if (!params.containsKey("arg")) { System.out.println("height <cm>"); return; }
                app.setHeight(Double.parseDouble(params.get("arg")));
                break;
            case "preweight":
                if (!params.containsKey("arg")) { System.out.println("preweight <kg>"); return; }
                app.setPreWeight(Double.parseDouble(params.get("arg")));
                break;
            case "add":
                if (!params.containsKey("arg")) { System.out.println("add <weight> [--date DATE] [--note TEXT]"); return; }
                double weight = Double.parseDouble(params.get("arg"));
                String date = params.get("date");
                String note = params.get("note");
                app.add(weight, date, note);
                break;
            case "list":
                app.list();
                break;
            case "stats":
                app.stats();
                break;
            case "bmi":
                app.showBMI();
                break;
            case "gain":
                app.showGain();
                break;
            case "trend":
                int days = params.containsKey("days") ? Integer.parseInt(params.get("days")) : 5;
                app.trend(days);
                break;
            case "export":
                String filename = params.getOrDefault("arg", "weights.csv");
                app.export(filename);
                break;
            default:
                System.out.println("Unknown command.");
        }
    }
}

File: pregnancy_weight.go
// pregnancy_weight.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
	"time"
)

type Entry struct {
	Date   string  `json:"date"`
	Weight float64 `json:"weight"`
	Note   string  `json:"note"`
	BMI    float64 `json:"bmi"`
}

type Data struct {
	Height    float64 `json:"height"`
	PreWeight float64 `json:"pre_weight"`
	Entries   []Entry `json:"entries"`
}

type App struct {
	data Data
	file string
}

func NewApp(file string) *App {
	app := &App{file: file}
	app.load()
	return app
}

func (a *App) load() {
	f, err := os.ReadFile(a.file)
	if err != nil {
		return
	}
	json.Unmarshal(f, &a.data)
}

func (a *App) save() {
	data, _ := json.MarshalIndent(a.data, "", "  ")
	os.WriteFile(a.file, data, 0644)
}

func (a *App) setHeight(cm float64) {
	a.data.Height = cm
	a.save()
	fmt.Printf("✅ Height set to %.1f cm\n", cm)
}

func (a *App) setPreWeight(kg float64) {
	a.data.PreWeight = kg
	a.save()
	fmt.Printf("✅ Pre‑pregnancy weight set to %.1f kg\n", kg)
}

func (a *App) bmi(weight float64) float64 {
	if a.data.Height <= 0 {
		return 0
	}
	h := a.data.Height / 100.0
	return weight / (h * h)
}

func (a *App) bmiCategory(bmiVal float64) string {
	if bmiVal < 18.5 {
		return "Underweight"
	} else if bmiVal < 25.0 {
		return "Normal"
	} else if bmiVal < 30.0 {
		return "Overweight"
	} else {
		return "Obese"
	}
}

func (a *App) add(weight float64, dateStr string, note string) {
	if a.data.Height <= 0 {
		fmt.Println("❌ Please set your height first: height <cm>")
		return
	}
	if dateStr == "" {
		dateStr = time.Now().Format("2006-01-02")
	}
	bmiVal := a.bmi(weight)
	entry := Entry{
		Date:   dateStr,
		Weight: weight,
		Note:   note,
		BMI:    bmiVal,
	}
	a.data.Entries = append(a.data.Entries, entry)
	sort.Slice(a.data.Entries, func(i, j int) bool {
		return a.data.Entries[i].Date < a.data.Entries[j].Date
	})
	a.save()
	fmt.Printf("✅ Logged %.1f kg on %s\n", weight, dateStr)
}

func (a *App) list() {
	if len(a.data.Entries) == 0 {
		fmt.Println("No entries.")
		return
	}
	fmt.Println("\n📋 Weight Entries:")
	for _, e := range a.data.Entries {
		note := ""
		if e.Note != "" {
			note = " – " + e.Note
		}
		bmiStr := ""
		if e.BMI > 0 {
			bmiStr = fmt.Sprintf(" (BMI: %.1f)", e.BMI)
		}
		fmt.Printf("  %s: %.1f kg%s%s\n", e.Date, e.Weight, bmiStr, note)
	}
}

func (a *App) stats() {
	if len(a.data.Entries) == 0 {
		fmt.Println("No entries.")
		return
	}
	weights := make([]float64, len(a.data.Entries))
	for i, e := range a.data.Entries {
		weights[i] = e.Weight
	}
	current := weights[len(weights)-1]
	total := len(weights)
	sum := 0.0
	min := weights[0]
	max := weights[0]
	for _, w := range weights {
		sum += w
		if w < min {
			min = w
		}
		if w > max {
			max = w
		}
	}
	avg := sum / float64(total)
	gain := 0.0
	if a.data.PreWeight > 0 {
		gain = current - a.data.PreWeight
	}
	fmt.Println("\n📊 Statistics:")
	fmt.Printf("  Entries: %d\n", total)
	fmt.Printf("  Current: %.1f kg\n", current)
	fmt.Printf("  Average: %.1f kg\n", avg)
	fmt.Printf("  Min: %.1f kg\n", min)
	fmt.Printf("  Max: %.1f kg\n", max)
	if a.data.PreWeight > 0 {
		fmt.Printf("  Total gain: %+.1f kg\n", gain)
	}
}

func (a *App) showBMI() {
	if len(a.data.Entries) == 0 {
		fmt.Println("No entries.")
		return
	}
	if a.data.Height <= 0 {
		fmt.Println("❌ Height not set.")
		return
	}
	current := a.data.Entries[len(a.data.Entries)-1].Weight
	bmiVal := a.bmi(current)
	cat := a.bmiCategory(bmiVal)
	fmt.Printf("\n📐 Current BMI: %.1f (%s)\n", bmiVal, cat)
	fmt.Println("BMI Range: 18.5 - 24.9")
}

func (a *App) showGain() {
	if len(a.data.Entries) == 0 {
		fmt.Println("No entries.")
		return
	}
	if a.data.PreWeight <= 0 {
		fmt.Println("❌ Pre‑pregnancy weight not set. Use: preweight <kg>")
		return
	}
	current := a.data.Entries[len(a.data.Entries)-1].Weight
	first := a.data.Entries[0].Weight
	gain := current - a.data.PreWeight
	fmt.Println("\n📈 Weight Gain:")
	fmt.Printf("  Pre‑pregnancy: %.1f kg\n", a.data.PreWeight)
	fmt.Printf("  First logged: %.1f kg\n", first)
	fmt.Printf("  Current: %.1f kg\n", current)
	fmt.Printf("  Total gain: %+.1f kg\n", gain)
}

func (a *App) trend(days int) {
	if len(a.data.Entries) < days {
		fmt.Printf("Need at least %d entries for trend.\n", days)
		return
	}
	recent := a.data.Entries[len(a.data.Entries)-days:]
	sum := 0.0
	for _, e := range recent {
		sum += e.Weight
	}
	avg := sum / float64(len(recent))
	fmt.Printf("\n📈 Trend (last %d days):\n", days)
	fmt.Printf("  Average weight: %.1f kg\n", avg)
	fmt.Println("  Recent entries:")
	start := len(recent) - 5
	if start < 0 {
		start = 0
	}
	for _, e := range recent[start:] {
		fmt.Printf("    %s: %.1f kg\n", e.Date, e.Weight)
	}
}

func (a *App) export(filename string) {
	if len(a.data.Entries) == 0 {
		fmt.Println("No entries.")
		return
	}
	f, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Error creating file: %v\n", err)
		return
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"Date", "Weight (kg)", "BMI", "Note"})
	for _, e := range a.data.Entries {
		bmiStr := ""
		if e.BMI > 0 {
			bmiStr = fmt.Sprintf("%.1f", e.BMI)
		}
		w.Write([]string{e.Date, fmt.Sprintf("%.1f", e.Weight), bmiStr, e.Note})
	}
	fmt.Printf("✅ Exported %d entries to %s\n", len(a.data.Entries), filename)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: pregnancy_weight <command> [options]")
		return
	}
	app := NewApp("pregnancy_weight.json")
	cmd := os.Args[1]

	switch cmd {
	case "height":
		if len(os.Args) < 3 {
			fmt.Println("height <cm>")
			return
		}
		cm, _ := strconv.ParseFloat(os.Args[2], 64)
		app.setHeight(cm)
	case "preweight":
		if len(os.Args) < 3 {
			fmt.Println("preweight <kg>")
			return
		}
		kg, _ := strconv.ParseFloat(os.Args[2], 64)
		app.setPreWeight(kg)
	case "add":
		if len(os.Args) < 3 {
			fmt.Println("add <weight> [--date DATE] [--note TEXT]")
			return
		}
		weight, _ := strconv.ParseFloat(os.Args[2], 64)
		dateStr := ""
		note := ""
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--date" && i+1 < len(os.Args) {
				dateStr = os.Args[i+1]
				i++
			}
			if os.Args[i] == "--note" && i+1 < len(os.Args) {
				note = os.Args[i+1]
				i++
			}
		}
		app.add(weight, dateStr, note)
	case "list":
		app.list()
	case "stats":
		app.stats()
	case "bmi":
		app.showBMI()
	case "gain":
		app.showGain()
	case "trend":
		days := 5
		for i := 2; i < len(os.Args); i++ {
			if os.Args[i] == "--days" && i+1 < len(os.Args) {
				days, _ = strconv.Atoi(os.Args[i+1])
				i++
			}
		}
		app.trend(days)
	case "export":
		filename := "weights.csv"
		if len(os.Args) > 2 {
			filename = os.Args[2]
		}
		app.export(filename)
	default:
		fmt.Println("Unknown command. Use height, preweight, add, list, stats, bmi, gain, trend, export.")
	}
}
